File: 703.kth-largest-element-in-a-stream.py
# @leet start
from heapq import heappush, heappop, heapify
class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.k = k

        heapify(nums)
        self.k_largest = nums

        while len(self.k_largest) > k:
            heappop(self.k_largest)

    def add(self, val: int) -> int:
        heappush(self.k_largest, val)

        if (len(self.k_largest) > self.k):
            heappop(self.k_largest)

        return self.k_largest[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)
# @leet end

File: 703.kth-largest-element-in-a-stream.cpp
// @leet start
#include <algorithm>
class KthLargest {
private:
    vector<int> heap;
    int k;
public:
    KthLargest(int k, vector<int>& nums) {
        this->k = k;
        for (int num : nums) {
            this->add(num);
        }
    }
    
    int add(int val) {
        if (this->heap.size() < this->k) {
            this->heap.push_back(val);
            std::push_heap(heap.begin(), heap.end(), std::greater<>{});
        } else {
            if (this->heap[0] < val) {
                std::pop_heap(heap.begin(), heap.end(), std::greater<>{});
                this->heap.pop_back();
                this->heap.push_back(val);
                std::push_heap(heap.begin(), heap.end(), std::greater<>{});
            }
        }
        return this->heap[0];
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */
// @leet end
